backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java

```diff
@@ -2003,4 +2005,14 @@ public Optional<CountryId> getBillToCountryId(@NonNull final InvoiceId invoiceId
 
 	}
 
+	@Override
+	public boolean hasInvoiceWithForexContract(@NonNull final OrderId orderId, @NonNull final ForexContractId contractId)
+	{
+		return getByOrderId(orderId)
+				.stream()
+				.map(InvoiceDAO::extractForeignContractRef)
+				.filter(Objects::nonNull)
+				.map(ForexContractRef::getForexContractId)
+				.anyMatch(contractId::equals);
```

Comment:
IMHO i find a bit extreme this kind of "decomposition" of an expression in Stream operations.

I would prefer to introduce and use a method called
```java
boolean isContractIdMatching(final I_C_Invoice invoice, ForexContractId expectedContractId)
```

or at least 
```java
.map(ForexContractRef::getForexContractId)
.anyMatch(contractId::equals);
```
could be written as
```java
.anyMatch(contractRef -> ForexContractId.equals(contractRef.getForexContractId(), contracted));
```

PS: i am open to discussions if needed

Reply:
refactored using second suggestion